### PR TITLE
Draft: initial windows support

### PR DESCRIPTION
**Pull request by ayushbindlish**
Hi @robert-lag 

I created the initial windows commit.
The Basic seems to work, but it requires more testing. Can you guide me as to what needs to be tested?
We can add cargo tests for the same.
Also, can we have a windows branch here against which I can raise windows PRs?

Thanks,
Ayush

### CODE REVIEW

**Comment by ayushbindlish** [comment]
It is not ready yet to merge in the main branch. @robert-lag

**Comment by robert-lag** [comment]
Sorry, I merged it just now without thinking (shouldn't program when I'm tired 😅) and reverted it afterwards.
 
> Can you guide me as to what needs to be tested?

Subcommands like `list`, `history`, `graph`,  `get-file-name` or `get-name` can simply be tested on the command line as they are just displaying information about the zettelkasten.

I think the following subcommands are most likely to break:

- `init` - Should create the `.zettelkasten` directory in the current directory or the one specified by the `-d` option and add the following files and folders:
    - cytoscape/
    - data.db
    - graph.js
    - history
    - index.html
    - note-template.md
    - style.css
- `add` - Should add the note, change the note name and update the database (./.zettelkasten/data.db).
- `rm` - Should remove the note and update the database (./.zettelkasten/data.db).
- `open` - Does opening the note and changing the note name work?
- `tui` - Does it display correctly? Do all functions work as you would expect?
    - Opening notes
    - Search using `/`
    - Add using `a`
    - Remove using `x`
- `-d` or `--dir` option - Sets the zettelkasten directory to something other than the current directory

Basically, I think there could be some errors when doing file operations (e.g. adding notes, removing notes, renaming notes). Additionally, it should be checked if the database (./.zettelkasten/data.db) has all notes and tags and connects them correctly (if you delete a tag in a note the corresponding entry in NoteTaggings should also disappear). But I find it unlikely that there will be any problems in the database.

I hope this helps. ;)

>  We can add cargo tests for the same.

I have never worked with cargo tests until now, but I will look into it. ;)

> Also, can we have a windows branch here against which I can raise windows PRs?

Sure, we can do that.

**Comment by ayushbindlish** [comment]
@robert-lag 
Thanks for the insights.

**Current status:**
- init works and creates the required files
- add works
- open works
- rm works
-  -d works
- history works
-  tui
    - a works but tui needs to be relaunched to see the updates
    - x works but tui needs to be relaunched to see the updates
    - / works
Except graph everything else seems to work fine on windows.
Need to test the same thing on Linux as well to make sure the current changes do not break anything.

**Current Issues:**
- Faced some library issues.
    - Had to install c++ build tools for windows. 
    - SQLite libraries were also downloaded and placed in the project's root directory.
I will look into build.rs or a separate script to automate fetching these dependencies.
- cargo build in debug mode complains of using unstable library features resulting in a failed build
- Only cargo build --release works


Did you face any similar setup issues on Linux? It would be good if we can do a readme for that. 

Also, environment variables like `EDITOR` and `directory` do not seem to be standard. Either we can use the defaults or ask the user to set these variables.

**Comment by robert-lag** [comment]
> a works but tui needs to be relaunched to see the updates

This issue is also present on linux. It should also be possible to refresh by pressing `Esc`. This will also send you back to the *start page*, where you can see your latest notes.

> Faced some library issues. 

I didn't have those issues on linux. But it could be that I already had all the necessary libraries installed from the beginning.

> I will look into build.rs or a separate script to automate fetching these dependencies.

Good idea.

> cargo build in debug mode complains of using unstable library features resulting in a failed build
> Only cargo build --release works

This seems weird to me. On linux both `cargo build` and `cargo build --release` work without problems.

> Also, environment variables like EDITOR and directory do not seem to be standard. Either we can use the defaults or ask the user to set these variables.

I don't know how other terminal apps handle this on Windows, but we could create a config file in `./.zettelkasten/` where these variables can be set. `brn` would basically look for configuration in the following order (if the first doesn't exist, it falls back to the next option):
1. `./.zettelkasten/config` (instead of this local config file a global one (in %APPDATA%) would also be possible)
2. Environment variables
3. Use default value

Or, like you said, we could just inform the user that these environment variables need to be set (perhaps in README.md). This option would be less work to implement. What do you think about it? Do you have any other ideas?
